### PR TITLE
Improve error on conflict for nix profile install

### PR DESCRIPTION
**Pull request by bobvanderlinden**
# Motivation

Currently `nix profile install` gives a confusing and unhelpful error message whenever a package is installed that has conflicting files. This often happens when installing a different version of the same package.

This PR is an attempt to improve the experience in such situations to some extend. I see this as a small improvement, but it doesn't really reach the end-goal of a good UX yet.

As an example, this is what happens before this PR:

```console
$ nix profile install github:cachix/devenv/v0.5
$ nix profile install github:cachix/devenv/v0.4
error: files '/nix/store/k0gwhiwfs8c218kph78vx75bph51iw31-devenv/bin/devenv' and '/nix/store/7i4hxhw44lr7r64zlmp4zq7cqpsx7smw-devenv/bin/devenv' have the same priority 5; use 'nix-env --set-flag priority NUMBER INSTALLED_PKGNAME' or type 'nix profile install --help' if using 'nix profile' to find out how to change the priority of one of the conflicting packages (0 being the highest priority)
```

This is what happens after this PR:

```console
$ nix profile install github:cachix/devenv/v0.5
$ nix profile install github:cachix/devenv/v0.4
error: There is an existing package that has a conflicting file:

         /nix/store/7i4hxhw44lr7r64zlmp4zq7cqpsx7smw-devenv/bin/devenv

       The conflicting file:

         /nix/store/k0gwhiwfs8c218kph78vx75bph51iw31-devenv/bin/devenv

       You can remove the existing package using:

         nix profile remove github:cachix/devenv/v0.4

       You can also install the new package next to the existing package by assigning a priority to the package.
       The conflicting packages have a priority of 5.
       You can prioritize the new package using:

         nix profile install github:cachix/devenv/v0.4 --priority 4

       You can prioritize the existing package using:

         nix profile install github:cachix/devenv/v0.4 --priority 6
```

The new changes allows `nix profile install` to give more context.

This also avoids mentions of `nix-env` and `nix profile` when using `builtins.buildEnv`. Since `buildEnv` can be used in other places (unrelated to the Nix profile), it makes sense to avoid mentioning `nix-env` and `nix profile`, as they likely do not apply in these cases.

# Context

Related to https://github.com/NixOS/nix/issues/7530

See [my suggestion](https://github.com/NixOS/nix/issues/7530#issuecomment-1419123524) for a good UX on `nix profile`s. The current PR is merely an improvement for the error message on conflicts.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

### CODE REVIEW

**Comment by domenkozar** [comment]
This looks great to me in terms of the error message improvement, but I'll let those that know the codebase comment on the C++ bits.

**Comment by bobvanderlinden** [comment]
I made some more changes so that the `INSTALLABLE` of the suggested `nix profile install` commands can be filled in. The implementation first throws an error containing the conflicting files. Then the package path (`/nix/store/XXXX-package`) is looked up for the conflicting files. Lastly the INSTALLABLE from the manifest is found using the package path. All of this information is used in the resulting error.

File: src/nix/profile.cc
Comment (asymmetric) [inline src/nix/profile.cc:370]: This reads confusing to me, as if the second store path is contained in the first.

```suggestion
                "An existing package already symlinks the following file to the user's profile:\n"
                "\n"
                "  %1%\n"
                "\n"
                "The conflicting file:\n"
                "\n"
                "  %3%\n"
```

File: src/libstore/path.hh
Comment (Ericson2314) [inline src/libstore/path.hh:74]: We don't want these, we want to use the parsing methods on `Store` instead.

File: src/libstore/builtins/buildenv.hh
Comment (Ericson2314) [inline src/libstore/builtins/buildenv.hh:34]: Side note, we really have a separate virtual method for *rendering* errors so we aren't storing both the structured data and a big long preformatted string. This is an awkard redundancy that just tempts people into overusing the plain `Error` type.

File: src/nix/profile.cc
Comment (Ericson2314) [inline src/nix/profile.cc:343]:
```suggestion
                auto maybeConflictingStorePath = store->maybeParseStorePath(conflictError.conflictingPkgPath);
```

File: src/libstore/path.hh
Comment (Ericson2314) [reply src/libstore/path.hh:74]:
```suggestion
```

File: src/libstore/path.cc
Comment (Ericson2314) [inline src/libstore/path.cc:71]:
```suggestion
```

**Comment by Ericson2314** [comment]
@bobvanderlinden fair warning I don't yet have perms to merge this, but I can still shepherd along making it an easier approval for those that do.

File: src/nix/profile.cc
Comment (bobvanderlinden) [reply src/nix/profile.cc:370]: Indeed, it's worded a bit convoluted. I'd like to avoid mentioning symlinking to the user. They don't necessarily know profiles are build from symlinks. I reworded it as:

> An existing package already provides the following file to the user's profile

File: src/nix/profile.cc
Comment (bobvanderlinden) [reply src/nix/profile.cc:343]: Oh, that is a lot more practical than introducing parseAbsolutePath :sweat_smile: Thanks :+1:

File: src/libstore/builtins/buildenv.hh
Comment (Ericson2314) [reply src/libstore/builtins/buildenv.hh:34]: Side note split into issue #7865.

Comment (bobvanderlinden) [reply src/libstore/builtins/buildenv.hh:34]: Hmm, I see I could override `what`, but quite a few places use `.msg()` instead of `what`. `msg` doesn't seem virtual.

I looked around in the codebase, but it seems there aren't that many custom (structured) errors, as they all use `MakeError`.

What would be the right way to implement 'lazy' error messages?

Comment (Ericson2314) [reply src/libstore/builtins/buildenv.hh:34]: @bobvanderlinden to be clear, I didn't mean this to block this PR. But if you would like to take a stab at this issue, I am happy to provide guidance over there.

**Comment by Ericson2314** [comment]
@bobvanderlinden This is starting to look quite good! Thanks for doing this!

Correct me if I am wrong, but `BuildEnvFileConflictError` is just used in a private-to-the-file `static` function, so you should be able to *not* expose it in the header, but just have it in `buildenv.cc`.

Since it is for internal use only, and will always be caught, I think it would be also fine to just make it inherit `std::exception` rather than `Error`, and then you don't have to write a nice message which would be dead code.

**Comment by bobvanderlinden** [comment]
> @bobvanderlinden This is starting to look quite good! Thanks for doing this!
> 
> Correct me if I am wrong, but `BuildEnvFileConflictError` is just used in a private-to-the-file `static` function, so you should be able to _not_ expose it in the header, but just have it in `buildenv.cc`.
> Since it is for internal use only, and will always be caught, I think it would be also fine to just make it inherit `std::exception` rather than `Error`, and then you don't have to write a nice message which would be dead code.

Ah you're right! That makes a lot of sense, thanks :+1: Updated the PR.

File: src/nix/profile.cc
Comment (Ericson2314) [inline src/nix/profile.cc:48]: I like the combinators! But try https://en.cppreference.com/w/cpp/container/set/contains or (if our C++20 is not new enough) https://en.cppreference.com/w/cpp/container/set/find

Comment (Ericson2314) [reply src/nix/profile.cc:48]: Then you don't need this method

File: src/libstore/builtins/buildenv.cc
Comment (Ericson2314) [inline src/libstore/builtins/buildenv.cc:16]:
```suggestion
/** This is private to this file and always caught, so we don't worry about it rendering in a nice fancy way etc. */
class BuildEnvFileConflictError : public std::exception
```

Comment (Ericson2314) [reply src/libstore/builtins/buildenv.cc:16]: And maybe line wrap it :)

**Comment by Ericson2314** [comment]
Thanks @bobvanderlinden, this is very close to being ready!

File: src/nix/profile.cc
Comment (bobvanderlinden) [reply src/nix/profile.cc]: I missed this method in my search for a right candidate. That makes it quite a bit cleaner :+1:

File: src/libstore/builtins/buildenv.cc
Comment (bobvanderlinden) [reply src/libstore/builtins/buildenv.cc:22]: I rewrote the comment a bit. Hope you like it ;)

**Comment by bobvanderlinden** [comment]
> Thanks @bobvanderlinden, this is very close to being ready!

Thanks for the quick reviews and suggestions. This improves my long forgotten C++ knowledge a bit too. :partying_face:

**Review by Ericson2314 — APPROVED** [review]
Looks good!

**Comment by Ericson2314** [comment]
Thanks so much, @bobvanderlinden, I've approved it and will recommend it for those with merge access.

**Comment by roberth** [comment]
Not testing error messages leads to a lot of untested code, including this code. We should consider testing more error messages. They're important too.

File: src/nix/profile.cc
Comment (asymmetric) [inline src/nix/profile.cc:362]: Here it's not 100% clear if "this" is referring to the one below or the one above, so I would reword as:

```suggestion
                "This is the conflicting file from the new package:\n"
```

**Comment by Ericson2314** [comment]
@roberth Agreed, see `src/libutil/tests/logging.cc` for an example of *commented out* :scream: testing of diagnostics. If we put the expected output in separate file (see "golden testing") these escaped string literals, it might be more maintainable.

**Comment by roberth** [comment]
> @roberth Agreed, see `src/libutil/tests/logging.cc` for an example of _commented out_ scream testing of diagnostics. If we put the expected output in separate file (see "golden testing") these escaped string literals, it might be more maintainable.

Crucially this needs a mode to overwrite the outputs, to fix the maintenance problem, turning tedious work into a skimmable diff review.

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting:

- @thufschmitt: this is a great improvement overall
- @edolstra: philosophical objection to putting manual-style content into error messages
  - better to have short description linking to the manual that tells how to fix something
  - @fricklerhandwerk: had this discussion in the past, currently don't have a good means of linking to the manual.
    - having versioned online documentation would make that easier, and should be a priority for @NixOS/documentation-team
   - then again, since the error message is adapted to the actual situation, this is something the manual can't do and therefore this change is fine
   - ideally we'd make this a markdown message and run a formatter on them for better presentation
- assigned to @fricklerhandwerk for reviewing the user-facing part

File: src/libstore/builtins/buildenv.cc
Comment (edolstra) [inline src/libstore/builtins/buildenv.cc:22]: Why is this a subclass of `std::exception` instead of `Error`? In most of the Nix code base we catch `Error`, not `std::exception`.

File: src/libstore/builtins/buildenv.cc
Comment (edolstra) [inline src/libstore/builtins/buildenv.cc:37]: It would be nicer to use designated initializers, e.g.
```
BuildEnvFileConflictError {
    .originalFile = ...,
    .conflictingFile = ...,
    .priority = ...
}
```
which is more readable and removes the need for a constructor.

File: src/libstore/builtins/buildenv.hh
Comment (edolstra) [inline src/libstore/builtins/buildenv.hh:30]: Idem, use designated initializers.

File: src/nix/profile.cc
Comment (edolstra) [inline src/nix/profile.cc:359]: Probably nicer to write this using multiline string literals, i.e. `R"(...)"`.

**Comment by bobvanderlinden** [comment]
> Not testing error messages leads to a lot of untested code, including this code. We should consider testing more error messages. They're important too.

Fully agree. I really like snapshot testing for these kinds of cases with large pieces of unstructured text. I have added a case to `tests/nix-profile.sh`. It's an example, but I don't know whether this is the way we want to test features like these. For now it works, but when something changes it might be useful to be able to update the snapshot automatically.

Maybe eventually look into a more reusable way to do snapshots? And/or combine it with a testing framework like [bats](https://github.com/bats-core/bats-core)?

Let me know whether the current test is wanted for this PR.

File: src/libstore/builtins/buildenv.cc
Comment (bobvanderlinden) [reply src/libstore/builtins/buildenv.cc:22]: See https://github.com/NixOS/nix/pull/7788#issuecomment-1438696402

File: src/libstore/builtins/buildenv.cc
Comment (bobvanderlinden) [reply src/libstore/builtins/buildenv.cc:37]: Hmm, I looked into this, is this even possible when it is inheriting from std::exception?

Not entirely sure, but it seems it is mostly possible for structs?

```
src/libstore/builtins/buildenv.cc:113:25: error: designated initializers cannot be used with a non-aggregate type ‘nix::BuildEnvFileConflictError’
```

File: src/nix/profile.cc
Comment (bobvanderlinden) [reply src/nix/profile.cc:359]: I have looked into various ways of doing multiline strings in C++. I think R"(...)" was one of them and it seemed it didn't support removing indentation, so it required me to put all text on the first column. The current solution seemed like one that did properly indent, which looked a bit better.

File: src/libstore/builtins/buildenv.cc
Comment (Ericson2314) [reply src/libstore/builtins/buildenv.cc:37]: Ah maybe `std::exception` isn't really needed, and then this can be a simple plain old struct.

File: tests/nix-profile.sh
Comment (Ericson2314) [inline tests/nix-profile.sh:147]: You'll need to detail with the store dir changing (use `$NIX_STORE_DIR`) and the hash change (not so sure).

Comment (bobvanderlinden) [reply tests/nix-profile.sh:147]: Yes, the hash is not reproducible on different machines with different test roots. I'm looking for a way to find the store path. I had `nix eval ${flake1Dir}"#default.outPath" --raw`, but it doesn't work yet. I'm going look into this at a later time.

File: src/nix/profile.cc
Comment (fricklerhandwerk) [inline src/nix/profile.cc:370]:
```suggestion
                "The new package can also be installed next to the existing one by assigning a different priority.\n"
```

I suppose it already has some priority, and the user would have to change it.

File: src/nix/profile.cc
Comment (fricklerhandwerk) [inline src/nix/profile.cc:358]:
```suggestion
                "An existing package already provides the following file:\n"
```

@edolstra seemed a bit surprised over the verbosity of the output, and I also think we should try to keep visual noise to a minimum while offering all the assistance we can. That's what I'm trying to do in these suggestions, so don't worry about getting many of them.

File: src/nix/profile.cc
Comment (fricklerhandwerk) [inline src/nix/profile.cc:366]:
```suggestion
                "To remove the existing package:\n"
```

Saying "you can ... using" doesn't really add information. The user always can something using something. We may leave it out. I understand addressing the user may sound friendlier, but a wall of text may be overwhelming, and should only happen if all of that is relevant information.

File: src/nix/profile.cc
Comment (fricklerhandwerk) [inline src/nix/profile.cc:372]:
```suggestion
                "To prioritise the new package:\n"
```

The docs team decided to use British English. We'll have to write it down somewhere...

File: src/nix/profile.cc
Comment (fricklerhandwerk) [inline src/nix/profile.cc:376]:
```suggestion
                "To prioritise the existing package:\n"
```

**Review by fricklerhandwerk — APPROVED** [review]
In really like the informative error message, we can use lots more of that! Only suggestions on more concise wording from my side. Great work.

File: tests/nix-profile.sh
Comment (bobvanderlinden) [reply tests/nix-profile.sh]: Ended up using `nix build --no-link --print-out-paths ${flake2Dir}"#default.out"`.

File: src/libstore/builtins/buildenv.cc
Comment (bobvanderlinden) [reply src/libstore/builtins/buildenv.cc]: Aah, yes, without using `std::exception` I could indeed change it to struct, which works for throwing :+1:

File: src/libstore/builtins/buildenv.hh
Comment (bobvanderlinden) [reply src/libstore/builtins/buildenv.hh:30]: By using struct this works. Incorporated into the PR :+1: Thanks for the suggestion, named-parameters-like is much clearer to read.

**Comment by bobvanderlinden** [comment]
The test did show that I made a mistake with the `nix profile remove` command where the conflicting installable was mixed up with the original installable, which I have now fixed. Good idea to add this test. Glad I did.

Thanks you all for the suggestions. I've incorporated most of them and resolved the discussions. Quite a few improvements to the PR already :+1:

> ideally we'd make this a markdown message and run a formatter on them for better presentation

Is this something I can look into for this PR?

**Comment by roberth** [comment]
> > ideally we'd make this a markdown message and run a formatter on them for better presentation
> 
> Is this something I can look into for this PR?

Better to do that in a followup, although it may make the testing story worse, as the layout becomes dependent on the length of the inputs.

**Comment by bobvanderlinden** [comment]
I have reworked the implementation in this PR and left the tests as-is.

I have removed `BuildProfileConflictError` and opted to merely throw `BuildEnvFileConflictError`. `BuildEnvFileConflictError` now holds `fileA` and `fileB`, which represent the 2 files that were conflicting inside `buildEnv`. Since we do not know at that time whether the files were from the original package or the newly installed package, I have called them A and B.

Determining which file belongs to the original/new package is done entirely in `CmdProfileInstall::run`. Packages that are being installed are placed at the end of manifest.elements, so from this I look up any package that matches one of the conflicting file paths at the end of `manifest.elements` to find which the new package and its conflicting file. I do the same thing for the already installed package, but by looking at the beginning of `manifest.elements`.

This gives an implementation that doesn't run into sorting issues that I can into before. Bonus: the code is cleaner this way as well.

**Comment by bobvanderlinden** [comment]
Hmm, MacOS in CI failed to build due to usage of C++20 ranges. I guess a different version of clang is used on MacOS to build Nix?
I now avoided using ranges and replaced it with `begin()` and `rbegin()` iterators. I tried compiling using clang and using `nix develop .#nix-gccStdenv` as well. :crossed_fingers:

**Comment by fricklerhandwerk** [comment]
@Ericson2314 I'll unsubscribe since the user-facing part is good, please continue with shepherding the implementation.

**Review by Ericson2314 — APPROVED** [review]

**Comment by Ericson2314** [comment]
Implementation looks good to me!

**Comment by fricklerhandwerk** [comment]
Merging it then.

**Comment by bobvanderlinden** [comment]
Thank you all 🥳👍

**Pull request merged**
